### PR TITLE
Improve test coverage and fix bugs for OData.Core UriParser feature

### PR DESCRIPTION
**Pull request by AlanWong-MS**
### Description
This change fills in test gaps and fixes bugs for priority items in the UriParser component of OData.Core. In this commit, `src/Microsoft.OData.Core/UriParser/Parsers/OpenTypeMethods.cs` is removed as dead code; a side note is that there is a similar class `OpenTypeMethods` defined and used in the test code.

### Checklist (Uncheck if it is not completed)

- [ x ] Test cases added
- [ x ] Build and test with one-click build and test script passed

### Additional work necessary
N/A

### CODE REVIEW

File: src/Microsoft.OData.Core/Parameterized.Microsoft.OData.Core.cs [hunk 1]
```diff
@@ -4374,6 +4374,27 @@ internal static class Strings {
...
+            get
+            {
+                return Microsoft.OData.TextRes.GetString(Microsoft.OData.TextRes.MetadataBinder_QueryOptionsBindStateCannotBeNull);
+            }
+        }
+
+        /// <summary>
+        /// A string like "The bind method cannot be null. In OData, the processing of query options should have a corresponding bind method."
+        /// </summary>
+        internal static string MetadataBinder_QueryOptionsBindMethodCannotBeNull {
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/Parameterized.Microsoft.OData.Core.cs]: nit: { on separate line

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/EntityIdFunctionalTests.cs [hunk 2]
```diff
@@ -21,6 +21,9 @@ public void ParseEntityId()
         {
             var parser = new ODataUriParser(HardCodedTestModel.TestModel, new Uri("http://example.org/"), new Uri("http://example.org/People(0)/$ref?$id=http://test.org/People(1)"));
             Assert.Equal(new Uri("http://test.org/People(1)"), parser.ParseEntityId().Id);
+
+            // Call PrseEntityId() again to test different code path
```
Comment (robward-ms) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/EntityIdFunctionalTests.cs]: >PrseEntityId [](start = 20, length = 12)

nit: ParseEntityId

Comment (robward-ms) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/EntityIdFunctionalTests.cs]: >test different code path [](start = 44, length = 24)

Not sure I understand how this tests a different code path.

File: src/Microsoft.OData.Core/UriParser/Parsers/OpenTypeMethods.cs [hunk 3]
```diff
@@ -1,731 +0,0 @@
...
-    using System.Diagnostics;
-    using System.Diagnostics.CodeAnalysis;
-    using System.Linq.Expressions;
-    using System.Reflection;
-    using Edm = Microsoft.OData.Edm;
-    #endregion
-
-    /// <summary>Use this class to perform late-bound operations on open properties.</summary>
-    /// <remarks>This class was copied from the product.</remarks>
-    internal static class OpenTypeMethods
```
Comment (robward-ms) [inline src/Microsoft.OData.Core/UriParser/Parsers/OpenTypeMethods.cs]: >OpenTypeMethods [](start = 26, length = 15)

can you include a comment in the commit/PR description as to why it's OK to delete this code?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/EntityIdFunctionalTests.cs (hunk 2 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/EntityIdFunctionalTests.cs]: Done.

Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/EntityIdFunctionalTests.cs]: The `ParseEntityId()` function first checks for the initialization of the object's entityIdSegment and based on our test coverage, we haven't hit the case in which the object has already been initialized. This second call is to simply hit that case where initialization has already taken place. I'll put that in the comment.

File: src/Microsoft.OData.Core/UriParser/Parsers/OpenTypeMethods.cs (hunk 3 above)
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/Parsers/OpenTypeMethods.cs]: Thanks for the catch. Updated description.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/Build.NetFramework/Microsoft.OData.Core.Tests.csproj [hunk 4]
```diff
@@ -437,9 +441,6 @@
       <Name>Microsoft.Test.OData.Utils</Name>
     </ProjectReference>
   </ItemGroup>
-  <ItemGroup>
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/Build.NetFramework/Microsoft.OData.Core.Tests.csproj]: why remove this one?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ExpressionLexerTests.cs [hunk 5]
```diff
@@ -85,6 +85,18 @@ public void ShouldReturnFalseIfNotLiteralToken()
             ExpressionTokenKind.Unknown.IsLiteralType().Should().BeFalse();
         }
 
+        [Fact]
+        public void CommaTokenIsKeyValueShouldReturnFalse()
+        {
+            CommaToken.IsKeyValueToken.Should().Be(false);
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ExpressionLexerTests.cs]: Be(false) ==> BeFalse()

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ExpressionLexerTests.cs [hunk 6]
```diff
@@ -85,6 +85,18 @@ public void ShouldReturnFalseIfNotLiteralToken()
...
+        [Fact]
+        public void CommaTokenIsKeyValueShouldReturnFalse()
+        {
+            CommaToken.IsKeyValueToken.Should().Be(false);
+        }
+
+        [Fact]
+        public void CommaTokenIsFunctionParameterTokenShouldReturnFalse()
+        {
+            CommaToken.IsFunctionParameterToken.Should().Be(false);
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ExpressionLexerTests.cs]: Same as here

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Parsers/UriPathParserTests.cs [hunk 7]
```diff
@@ -263,6 +263,40 @@ public void ParsePathIgnoresQueryString()
             });
         }
 
+        [Fact]
+        public void IsCharHexDigitTest()
+        {
+            UriParserHelper.IsCharHexDigit(' ').Should().Be(false);
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Parsers/UriPathParserTests.cs]: Be(false) ==> BeFalse()

Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Parsers/UriPathParserTests.cs]: same as others

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/UriEdmHelpersTests.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            result.Should().Be(segment.NavigationProperty);
+        }
+
+        [Fact]
+        public void GetMostDerivedTypeFromPathWithMetadataSegmentAndNullReturnsType()
+        {
+            var segment = MetadataSegment.Instance;
+            ODataPath odataPath = new ODataPath(segment);
+            var result = UriEdmHelpers.GetMostDerivedTypeFromPath(odataPath, null);
+            result.Should().Be(null);
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/UriEdmHelpersTests.cs]: Be(null) ==> BeNull()

File: test/FunctionalTests/Microsoft.OData.Core.Tests/Build.NetFramework/Microsoft.OData.Core.Tests.csproj (hunk 4 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/Build.NetFramework/Microsoft.OData.Core.Tests.csproj]: This should not be removed. I think this was mismatched during one of my rebases. Thanks for the catch!

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/UriEdmHelpersTests.cs (hunk 8 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/UriEdmHelpersTests.cs]: Done.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Parsers/UriPathParserTests.cs (hunk 7 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Parsers/UriPathParserTests.cs]: Done.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ExpressionLexerTests.cs (hunk 6 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ExpressionLexerTests.cs]: Done.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ExpressionLexerTests.cs (hunk 5 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/ExpressionLexerTests.cs]: Done.

**Comment by robward-ms** [comment]
Please Squash to merge a single commit.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using FluentAssertions;
+using Microsoft.OData.UriParser;
+using System;
+using System.Collections.Generic;
+using Xunit;
+
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs]: remove unused the white space line

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+{
+    /// <summary>
+    /// Unit test for the MetadataBinding class.
+    /// </summary>
+    public class MetadataBindingTests
+    {
+        private static readonly ODataUriParserConfiguration configuration = new ODataUriParserConfiguration(HardCodedTestModel.TestModel);
+
+        public MetadataBindingTests()
+        {
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs]: Can we remove this empty constructor?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+            long? result = MetadataBinder.ProcessSkip(1);
+            result.Should().Be(1);
+        }
+
+        [Fact]
+        public void MetadataBinderProcessTopWithNullReturnsNull()
+        {
+            long? result = MetadataBinder.ProcessTop(null);
+            result.Should().Be(null);
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs]: Is there a function named "BeNull()"?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        {
+            BindInfo bindInfo = new BindInfo();
+
+            Action metadataBinderAction = () => MetadataBinder.ProcessQueryOptions(bindInfo.BindingState, null);
+            metadataBinderAction.ShouldThrow<ODataException>(
+                ODataErrorStrings.MetadataBinder_QueryOptionsBindMethodCannotBeNull);
+        }
+
+        [Fact]
+        public void MetadataBinderProcessQueryOptionsWithBindMethodReturnNullShouldReturnEmptyList()
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs]: .... ReturnNullReturnEmptyList()

Double Return validation?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs (hunk 9 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs]: Done.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs (hunk 11 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs]: FluentAssertions appears to not have defined the BeNull() extension for numbers. I've updated other areas to BeNull() though.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs (hunk 12 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs]: It's the name of a delegate so there's not actually double return validation. I've updated the name of the test for better clarity.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs (hunk 10 above)
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingTests.cs]: Done.

**Pull request merged**